Mathlib/MeasureTheory/Integral/FiniteMeasureCharFun.lean

```diff
@@ -0,0 +1,223 @@
+
+`fun w => ∫ v, e (-L v w) ∂P = ∫ v, probChar w ∂P`
+
+We show:
+- `MeasureTheory.ext_of_charFun_eq`: If the characteristic functions of two finite measures `P`
+  and `P'` are equal, then `P = P'`. In other words, characteristic functions separate finite
+  measures.
+-/
+
+open MeasureTheory Filter BoundedContinuousFunction Complex
+
+section ProbChar
+
+variable {V W : Type*} [AddCommGroup V] [Module ℝ V] [TopologicalSpace V]
+    [AddCommGroup W] [Module ℝ W] [TopologicalSpace W]
+    {e : AddChar ℝ Circle} {L : V →ₗ[ℝ] W →ₗ[ℝ] ℝ}
+    {he : Continuous e} {hL : Continuous fun p : V × W ↦ L p.1 p.2}
+
+/-- The bounded continuous mapping `fun v ↦ e (L v (Multiplicative.toAdd w))` from `V` to `ℂ`. -/
+def probChar (he : Continuous e) (hL : Continuous fun p : V × W ↦ L p.1 p.2)
+    (w : Multiplicative W) :
+    V →ᵇ ℂ where
+  toFun := fun v ↦ e (L v (Multiplicative.toAdd w))
+  continuous_toFun :=
+    continuous_induced_dom.comp (he.comp (hL.comp (Continuous.prodMk_left w)))
+  map_bounded' := by
+    refine ⟨2, fun x y ↦ ?_⟩
+    calc dist _ _
+      ≤ (‖_‖ : ℝ) + ‖_‖ := dist_le_norm_add_norm _ _
+    _ ≤ 1 + 1 := add_le_add (by simp) (by simp)
+    _ = 2 := by ring
+
+@[simp]
+lemma probChar_apply (w : Multiplicative W) (v : V) :
+    probChar he hL w v = e (L v (Multiplicative.toAdd w)) := rfl
+
+@[simp]
+lemma probChar_one : probChar he hL 1 = 1 := by ext; simp
+
+lemma probChar_mul (x y : Multiplicative W) :
+    probChar he hL (x * y) = probChar he hL x * probChar he hL y := by
+  ext
+  simp [e.map_add_eq_mul]
+
+lemma probChar_inv (w : Multiplicative W) :
+    probChar he hL w⁻¹ = star (probChar he hL w) := by ext; simp
+
+theorem probChar_SeparatesPoints (he : Continuous e) (he' : e ≠ 1)
+    (hL : Continuous fun p : V × W ↦ L p.1 p.2) (hL' : ∀ v ≠ 0, L v ≠ 0) {v v' : V} (hv : v ≠ v') :
+    ∃ w : W, probChar he hL w v ≠ probChar he hL w v' := by
+  obtain ⟨w, hw⟩ := DFunLike.ne_iff.mp (hL' (v - v') (sub_ne_zero_of_ne hv))
+  obtain ⟨a, ha⟩ := DFunLike.ne_iff.mp he'
+  use (a / (L (v - v') w)) • w
+  simp only [map_sub, LinearMap.sub_apply, probChar_apply, ne_eq]
+  rw [← div_eq_one_iff_eq (Circle.coe_ne_zero _), div_eq_inv_mul, ← coe_inv_unitSphere,
+    ← e.map_neg_eq_inv, ← Submonoid.coe_mul, ← e.map_add_eq_mul, OneMemClass.coe_eq_one]
+  calc e (- L v' ((a / (L v w - L v' w)) • w) + L v ((a / (L v w - L v' w)) • w))
+  _ = e (- (a / (L v w - L v' w)) • L v' w + (a / (L v w - L v' w)) • L v w) := by
+    congr
+    · rw [neg_smul, ← map_smul (L v')]
+    · rw [← map_smul (L v)]
+  _ = e ((a / (L (v - v') w)) • (L (v - v') w)) := by
+    simp only [neg_mul, map_sub, LinearMap.sub_apply]
+    congr
+    module
+  _ = e a := by
+    congr
+    simp only [map_sub, LinearMap.sub_apply, smul_eq_mul]
+    rw [div_mul_cancel₀]
+    convert hw
+    simp
+  _ ≠ 1 := ha
+
+/-- Monoid homomorphism mapping `w` to `fun v ↦ e (L v (Multiplicative.toAdd w))`. -/
+def expInnerMulI (he : Continuous e) (hL : Continuous fun p : V × W ↦ L p.1 p.2) :
```

Comment:
The name `expInnerMulI` made sense for an old version of my code where `e` was the exponential of an inner product multiplied with `I`, but it does not really make sense here for a general `e`.
Same remark for `expInnerMulIₐ` or even `expPoly`.